### ISSUE TITLE
Application type mismatch complains about convertible types

### ISSUE THREAD

**Issue by JasonGross** [leanprover/lean4#343]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover/lean4/issues).
    * Reduced the issue to a self-contained, reproducible test case.

### Description

```lean
structure CatIsh where
  Obj : Type o
  Hom : Obj → Obj → Type m
infixr:75 " ~> " => (CatIsh.Hom _)
structure FunctorIsh (C D : CatIsh) where
  onObj : C.Obj → D.Obj
  onHom : ∀ {s d : C.Obj}, (s ~> d) → (onObj s ~> onObj d)
def Catish : CatIsh :=
  {
    Obj := CatIsh
    Hom := FunctorIsh
  }
universes m o
unif_hint (mvar : CatIsh) where
  Catish.{m,o} =?= mvar |- mvar.Obj =?= CatIsh.{m,o}
structure CtxSyntaxLayerParamsObj where
  Ct : CatIsh
def CtxSyntaxLayerParams : CatIsh :=
  {
    Obj := CtxSyntaxLayerParamsObj
    Hom := sorry
  }
def CtxSyntaxLayerTy := CtxSyntaxLayerParams ~> Catish
/-
23:5: application type mismatch
  CatIsh.Hom Catish CtxSyntaxLayerParams Catish
argument has type
  CatIsh
but function has type
  Catish.Obj → Type (max (max u_1 u_2) (u_3 + 1) (u_4 + 1))
-/
```

However, `Catish.Obj` and `CatIsh` are convertible.  Moreover, if I replace `Ct : CatIsh` in `structure CtxSyntaxLayerParamsObj` with `Ct : Type _`, then the definition passes.

### Versions

```
$ lean --version
Lean (version 4.0.0-nightly-2021-03-10, commit 4c82f5c68888, Release)
$ uname -a
Linux JasonGross-X1 4.19.128-microsoft-standard #1 SMP Tue Jun 23 12:58:10 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.2 LTS
Release:        20.04
Codename:       focal
$ cmd.exe /C ver
'\\wsl$\Ubuntu\home\jgross\Documents\GitHub\syntax'
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.

Microsoft Windows [Version 10.0.19041.804]
```

**Comment by leodemoura** [comment]
After the bug fix https://github.com/leanprover/lean4/commit/847f95021aea1794f3569723bda46c13c3eb0ff1 
The error message is
```
error: stuck at solving universe constraints
  max (?u.959+1) (?u.958+1) =?= max (?u.965+1) (?u.964+1)
```
The situation is similar to issue #342. I don't see a solution unless we approximate and solve the constraint using
```
 ?u.959+1 =?= ?u.965+1
 ?u.958+1 =?= ?u.964+1
```

**Comment by JasonGross** [comment]
I think the error should say where this constraint came from, something like
```
23:5: application type mismatch
  CatIsh.Hom Catish CtxSyntaxLayerParams Catish
first argument
  Catish
has type
  CatIsh.{whatever universes}
but is expected to have type
  Catish.Obj{whatever universes}
stuck at solving universe constraints
  max (?u.959+1) (?u.958+1) =?= max (?u.965+1) (?u.964+1)
```
or something like that.  Once the error message is sufficiently informative, I'm happy for this to be closed and leave #342 to be about what to do with universes
